internal/venice-common/src/main/java/com/linkedin/venice/pubsub/manager/TopicManager.java

```diff
@@ -651,72 +618,110 @@ public synchronized boolean containsTopicAndAllPartitionsAreOnline(
   }
 
-  public long getProducerTimestampOfLastDataRecord(PubSubTopicPartition pubSubTopicPartition, int retries) {
-    return partitionOffsetFetcher.getProducerTimestampOfLastDataRecord(pubSubTopicPartition, retries);
+  public long getLatestOffsetCached(PubSubTopic pubSubTopic, int partitionId) {
+    return topicMetadataFetcher.getLatestOffsetCached(new PubSubTopicPartitionImpl(pubSubTopic, partitionId));
+  }
+
+  public long getProducerTimestampOfLastDataMessageWithRetries(PubSubTopicPartition pubSubTopicPartition, int retries) {
+    return topicMetadataFetcher.getProducerTimestampOfLastDataMessageWithRetries(pubSubTopicPartition, retries);
+  }
+
+  public long getProducerTimestampOfLastDataMessageCached(PubSubTopicPartition pubSubTopicPartition) {
+    return topicMetadataFetcher.getProducerTimestampOfLastDataMessageCached(pubSubTopicPartition);
   }
 
   /**
    * Get offsets for only one partition with a specific timestamp.
    */
-  public long getPartitionOffsetByTime(PubSubTopicPartition pubSubTopicPartition, long timestamp) {
-    return partitionOffsetFetcher.getPartitionOffsetByTime(pubSubTopicPartition, timestamp);
+  public long getOffsetByTime(PubSubTopicPartition pubSubTopicPartition, long timestamp) {
+    return topicMetadataFetcher.getOffsetForTimeWithRetries(pubSubTopicPartition, timestamp, 25);
   }
 
   /**
-   * Get a list of {@link PubSubTopicPartitionInfo} objects for the specified topic.
-   * @param topic
-   * @return
+   * Invalidate the cache for the given topic and all its partitions.
+   * @param pubSubTopic the topic to invalidate
    */
-  public List<PubSubTopicPartitionInfo> partitionsFor(PubSubTopic topic) {
-    return partitionOffsetFetcher.partitionsFor(topic);
+  public CompletableFuture<Void> invalidateCache(PubSubTopic pubSubTopic) {
+    return topicMetadataFetcher.invalidateKeyAsync(pubSubTopic);
   }
 
-  public String getPubSubBootstrapServers() {
-    return this.pubSubBootstrapServers;
+  /**
+   * Prefetch and cache the latest offset for the given topic-partition.
+   * @param pubSubTopicPartition the topic-partition to prefetch and cache the latest offset for
+   */
+  public void prefetchAndCacheLatestOffset(PubSubTopicPartition pubSubTopicPartition) {
+    topicMetadataFetcher.populateCacheWithLatestOffset(pubSubTopicPartition);
   }
 
-  @Override
-  public synchronized void close() {
-    Utils.closeQuietlyWithErrorLogged(partitionOffsetFetcher);
-    pubSubReadOnlyAdminAdapter.ifPresent(Utils::closeQuietlyWithErrorLogged);
-    pubSubWriteOnlyAdminAdapter.ifPresent(Utils::closeQuietlyWithErrorLogged);
+  public String getPubSubClusterAddress() {
+    return this.pubSubClusterAddress;
   }
 
-  // For testing only
-  public void setTopicConfigCache(Cache<PubSubTopic, PubSubTopicConfiguration> topicConfigCache) {
-    this.topicConfigCache = topicConfigCache;
+  @Override
+  public void close() {
+    if (isClosed.get()) {
+      logger.warn("{} is already closed", this);
```

Comment:
`return` directly if the TopicManager is already closed?

Reply:
Oh yeah I missed return here. Thanks

Reply:
done